src/common/libflux/future.c

```diff
@@ -673,10 +674,17 @@ void flux_future_fatal_error (flux_future_t *f, int errnum, const char *errstr)
 const char *flux_future_error_string (flux_future_t *f)
 {
     if (f) {
-        if (f->fatal_errnum_valid)
-            return f->fatal_errnum_string;
-        if (f->result_valid)
-            return f->result.errnum_string;
+        if (f->fatal_errnum_valid) {
+            if (f->fatal_errnum_string)
+                return f->fatal_errnum_string;
+            return flux_strerror (f->fatal_errnum);
+        }
+        else if (f->result_valid) {
+            if (f->result.errnum_string)
+                return f->result.errnum_string;
+            return flux_strerror (f->result.errnum);
```

Comment:
Is this right?  Is `f->result.errnum` valid on this leg?  Or `f->result.errnum_string` for that matter?

Should this leg just return "Success" to  be clearer?

Reply:
Is it possible you're thinking `result_valid` to mean the future fulfillment was a non-error fulfillment?  `result_valid` actually means there was fulfillment that is "legit", but it may or may not be an errored fulfillment.  If it's a non-error fulfillment, `errnum_string` should always be NULL and `errnum` should be 0.  Thus, `flux_strerror (0)` returns success.

Reply:
Can it be a non-error fulfillment?  It's in the else leg of a test on `fatal_errnum_valid`...

Reply:
Also, that function should never return NULL if it's going to be used unconditionally in error paths.  If it manages to be called when a future is not fulfilled, it should probably return a string like "future is not fulfilled".

Reply:
> Edit: I mean, can that block be anything but a non-error fulfillment?

The `if (f->result_valid)` block?  It can be a error or non-error fulfillment.  If it's a non-error, the `f->result.errnum` is 0, leading to return of "Success" as a string.

Reply:
This probably just needs an extra comment like
```C
/*  Future does not have a fatal error, but may have a valid error or non-error result */
```
I think the "fatal_errnum" was confusing me for a bit as well...